subsys/logging/log_backend_fs.c

```diff
@@ -0,0 +1,263 @@
+		/* If autostart, mount file system automatically. */
+		struct fs_mount_t *mp = &lfs_storage_mnt;
+
+		rc = fs_mount(mp);
+		if (rc < 0) {
+			backend_state = BACKEND_FS_CORRUPTED;
+
+			return;
+		}
+	}
+
+	rc = allocate_new_file(&file);
+	backend_state = (rc ? BACKEND_FS_CORRUPTED : BACKEND_FS_OK);
+}
+
+static void panic(struct log_backend const *const backend)
+{
+	/* In case of panic deinitialize backend. It is better to keep
+	 * current data rather than log new and risk of failure.
+	 */
+	log_backend_deactivate(backend);
+}
+
+static void dropped(const struct log_backend *const backend, uint32_t cnt)
+{
+	ARG_UNUSED(backend);
+
+	log_backend_std_dropped(&log_output, cnt);
+}
+
+static int allocate_new_file(struct fs_file_t *file)
+{
+	/* In case of no log file or current file fills up
+	 * create new log file.
+	 */
+	int rc;
+	int file_num = 0;
+	struct fs_statvfs stat;
+
+	assert(file);
+
+	if (backend_state == BACKEND_FS_NOT_INITIALIZED) {
+		/* Search for the last used log number. */
+		struct fs_dir_t dir = { 0 };
+
+		rc = fs_opendir(&dir, CONFIG_LOG_BACKEND_FS_MOUNT_POINT);
+
+		while (rc >= 0) {
+			struct fs_dirent ent = { 0 };
+
+			rc = fs_readdir(&dir, &ent);
+			if ((rc < 0) || (ent.name[0] == 0)) {
+				break;
+			}
+			if (strstr(ent.name, CONFIG_LOG_BACKEND_FS_FILE_PREFIX)
+			    != NULL) {
+				++file_ctr;
+				file_num = atoi(ent.name + LOG_PREFIX_LEN) + 1;
+			}
+		}
+
+		(void)fs_closedir(&dir);
+		if (rc < 0) {
+			goto out;
+		}
+
+		backend_state = BACKEND_FS_OK;
+	} else {
+		fs_close(file);
+		char *name = strstr(fname, CONFIG_LOG_BACKEND_FS_FILE_PREFIX);
+
+		file_num = atoi(name + LOG_PREFIX_LEN) + 1;
+	}
+
+	sprintf(fname, "%s/%s%04d", CONFIG_LOG_BACKEND_FS_MOUNT_POINT,
+		CONFIG_LOG_BACKEND_FS_FILE_PREFIX, file_num);
+	rc = fs_statvfs(CONFIG_LOG_BACKEND_FS_MOUNT_POINT, &stat);
+	if (rc < 0) {
+		goto out;
+	}
+
+	/* Check if there is enough space to write file or max files number
+	 * is not exceeded.
+	 */
+	while ((file_ctr >= CONFIG_LOG_BACKEND_FS_FILES_LIMIT) ||
+	       ((stat.f_bfree * stat.f_frsize) <=
+		CONFIG_LOG_BACKEND_FS_FILE_SIZE)) {
+
+		if (IS_ENABLED(CONFIG_LOG_BACKEND_FS_OVERWRITE)) {
+
+			rc = del_oldest_log();
+			if ((rc < 0) && (rc != -ENOENT)) {
+				goto out;
+			}
+			rc = fs_statvfs(CONFIG_LOG_BACKEND_FS_MOUNT_POINT,
+					&stat);
+			if (rc < 0) {
+				goto out;
+			}
+
+		} else {
+			return -ENOSPC;
+		}
+	}
+
+	rc = fs_open(file, fname, FS_O_CREATE | FS_O_WRITE);
+	if (rc < 0) {
+		goto out;
+	}
+	++file_ctr;
+
+out:
+	return rc;
+}
+
+static int del_oldest_log(void)
+{
+	struct fs_dir_t dir = { 0 };
+	int rc;
+	char del_file_path[MAX_FILE_NAME] = "\0";
+
+	/* Look for a first file with log prefix. */
+	rc = fs_opendir(&dir, CONFIG_LOG_BACKEND_FS_MOUNT_POINT);
+	while (rc >= 0) {
+		struct fs_dirent ent = { 0 };
+
+		rc = fs_readdir(&dir, &ent);
+		if (rc < 0) {
+			break;
+		}
+		if (strstr(ent.name, CONFIG_LOG_BACKEND_FS_FILE_PREFIX) !=
```

Comment:
This seems fragile.  What guarantee is there that the dirent order is oldest first?  This should extract the sequence number from the file name and use the numerically smallest one found.

It should also deal with sequence number wraparound if that's going to be handled.